### PR TITLE
Add support for Sphinx documentation builder

### PR DESCRIPTION
**Pull request by dozzie**

### CODE REVIEW

**Comment by essen** [comment]
Looks fine so far. Would be good to have a few tests, I'm not the most familiar with Sphinx and it'd show me how it's supposed to work.

**Comment by essen** [comment]
All good I think, it works here also. It's however missing an entry in build.config to include it in Erlang.mk at build time, and of course a documentation chapter depending on whether this plugin should be experimental or is ready for general use.

**Comment by dozzie** [comment]
I added a chapter to the documentation. I hope it's good enough.

File: doc/src/guide/sphinx.asciidoc [hunk 1]
```diff
@@ -0,0 +1,134 @@
+[[sphinx]]
+== Sphinx documentation
+
+Erlang.mk includes targets for running
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: for running the

File: doc/src/guide/sphinx.asciidoc [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+http://www.sphinx-doc.org/[Sphinx documentation generator], which can produce
+documentation in various formats, like HTML, man pages, Texinfo, LaTeX, and
+others.
+
+=== Writing Sphinx documentation
+
+Sphinx generates documentation form a set of
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: from

File: doc/src/guide/sphinx.asciidoc [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+http://www.sphinx-doc.org/en/stable/rest.html[reST] documents. There is
+a http://www.sphinx-doc.org/en/stable/tutorial.html[quick start guide] on
+Sphinx' website. For Erlang.mk, we'll set up a minimal environment instead.
+
+=== Basic setup
+
+By default, Erlang.mk expects Sphinx documentation to be placed in 'doc'
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: in the 'doc'

File: doc/src/guide/sphinx.asciidoc [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+[literal]
+----
+Other Page
+==========
+
+Lorem ipsum dolor sit amet...
+----
+
+The files above are enough to build HTML documentation to 'html' directory.
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: to the 'html'

File: doc/src/guide/sphinx.asciidoc [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+[source,bash]
+$ make docs     # all the docs, including EDoc and AsciiDoc if applicable
+$ make sphinx   # Sphinx docs specifically
+
+=== Erlang.mk configuration
+
+Default Erlang.mk settings are equivalent to adding to project's makefile
+following lines:
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: Erlang.mk defaults to the following configuration:

File: doc/src/guide/sphinx.asciidoc [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+[source,make]
+SPHINX_FORMATS = html
+SPHINX_SOURCE = doc
+
+To change the location of Sphinx sources, you need to set `$(SPHINX_SOURCE)`
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: to set the

File: doc/src/guide/sphinx.asciidoc [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+variable. 'conf.py' file should also be placed in that directory, unless you
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: The 'conf.py' file

File: doc/src/guide/sphinx.asciidoc [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+specify `$(SPHINX_CONFDIR)`.
+
+Variable `$(SPHINX_OPTS)` allows to provide options to `sphinx-build`, which
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: The variable

File: doc/src/guide/sphinx.asciidoc [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+is particularly useful for `-D name=value` options. You can even forego
+'doc/conf.py' file, using `-D name=value` in combination with `-C` option,
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: with the

File: doc/src/guide/sphinx.asciidoc [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+though you will need to manually call `make sphinx` or add `sphinx` target to
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: or add the

File: doc/src/guide/sphinx.asciidoc [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+dependencies of `docs`.
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: To be honest best say: or add `docs:: sphinx` to your Makefile.

But I thought from `$ make docs     # all the docs, including EDoc and AsciiDoc if applicable` that sphinx would run. If it doesn't, then it's not worth mentioning `make docs` in this chapter.

File: doc/src/guide/sphinx.asciidoc [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+
+`$(SPHINX_FORMATS)` variable lists formats to generate. By default only HTML
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: The ... variable

File: doc/src/guide/sphinx.asciidoc [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+is generated, but it can also include man pages or LaTeX document for building
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: but it can also build

LaTeX documents

File: doc/src/guide/sphinx.asciidoc [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+PDF. See
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: See the [description of the ...]

File: doc/src/guide/sphinx.asciidoc [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+http://www.sphinx-doc.org/en/stable/invocation.html#cmdoption-sphinx-build-b[description of `-b` option]
+for `sphinx-build` for list of known formats.
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: for a list

File: doc/src/guide/sphinx.asciidoc [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+
+Formats are by default generated to a directory called after the format
+('html' for HTML, 'man' for man pages, and so on). To change this behaviour
+for specific format, you can set `$(sphinx_$(format)_output)` variable, e.g.
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: for a specific format, you can set the

File: doc/src/guide/sphinx.asciidoc [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+`$(sphinx_html_output)` for 'html' or `$(sphinx_man_output)` for 'man'.
+There are also `$(sphinx_$(format)_opts)` variables for setting `sphinx-build`
+options for a single format only.
+
+=== Generating man pages
+
+To generate man pages, you need to include `man` in `$(SPHINX_FORMATS)` in
+your makefile and define `man_pages` option in 'doc/conf.py':
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: Makefile (for consistency)

and define the ... option

File: doc/src/guide/sphinx.asciidoc [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+
+[source,python]
+----
+man_pages = [
+    ('doc_name', 'page_name', 'Manpage Title', ['Page Author'], 1),
+]
+----
+
+As http://www.sphinx-doc.org/en/stable/config.html#options-for-manual-page-output[Sphinx documentation]
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: As the

File: doc/src/guide/sphinx.asciidoc [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+says, the structure is following:
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: the structure is:

"following" is not necessary

File: doc/src/guide/sphinx.asciidoc [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+  generated man page on a position that's used by `apropos` command
+* 'Page Author' (or more of them) will be included in autogenerated 'AUTHOR'
+  section, and leaving this field empty disables generating 'AUTHOR' section
+* '1' is the number of section of the man page
+
+With the above configuration (and Erlang.mk defaults), 'doc/doc_name.rst' will
+be used to generate 'man/page_name.1'.
+
+NOTE: You probably want to include a link to the man page in other
+documentation, possibly in 'doc/index.rst'.
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: I don't understand this note.

File: doc/src/guide/sphinx.asciidoc [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+----
+man_pages = [
+    ('doc_name', 'page_name', 'Manpage Title', ['Page Author'], 1),
+]
+----
+
+As http://www.sphinx-doc.org/en/stable/config.html#options-for-manual-page-output[Sphinx documentation]
+says, the structure is following:
+
+* 'doc_name' is the path to man page's source (relative `$(SPHINX_SOURCE)`),
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: to the

File: doc/src/guide/sphinx.asciidoc [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+  without the '.rst' suffix
+* 'page_name' is the name of resulting man page, which will be used as a base
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: of the resulting

File: doc/src/guide/sphinx.asciidoc [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+  for output file name and will be included in generated man page
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: for the

in the generated

File: doc/src/guide/sphinx.asciidoc [hunk 24]
```diff
@@ -0,0 +1,134 @@
...
+* 'Manpage Title' is a short, one-line description, which will be included in
+  generated man page on a position that's used by `apropos` command
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: in the generated

by the ... command

File: doc/src/guide/sphinx.asciidoc [hunk 25]
```diff
@@ -0,0 +1,134 @@
...
+* 'Page Author' (or more of them) will be included in autogenerated 'AUTHOR'
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: in the autogenerated

File: doc/src/guide/sphinx.asciidoc [hunk 26]
```diff
@@ -0,0 +1,134 @@
...
+  section, and leaving this field empty disables generating 'AUTHOR' section
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: section. Leaving this

generated the

File: doc/src/guide/sphinx.asciidoc [hunk 27]
```diff
@@ -0,0 +1,134 @@
...
+* '1' is the number of section of the man page
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: number of the man page section

File: doc/src/guide/sphinx.asciidoc [hunk 28]
```diff
@@ -0,0 +1,134 @@
...
+
+With the above configuration (and Erlang.mk defaults), 'doc/doc_name.rst' will
```
Comment (essen) [inline doc/src/guide/sphinx.asciidoc]: Erlang.mk's

**Comment by essen** [comment]
I have time today I can do the changes and merge this afternoon if you can't before that.

File: doc/src/guide/sphinx.asciidoc (hunk 11 above)
Comment (essen) [reply doc/src/guide/sphinx.asciidoc]: Nevermind, I understand now.

**Comment by essen** [comment]
Merged, thanks!

**Pull request closed without merging**